serial_darwin.go

```diff
@@ -0,0 +1,94 @@
+// +build darwin
+
+package serial
+
+import (
+	"io"
+	"os"
+	"syscall"
+	"unsafe"
+)
+
+// Darwin specific IOCTL constants
+const (
+	IOSSIOSPEED = 0x80045402 // _IOW('T', 2, speed_t)
+)
+
+func openPort(name string, baud int) (rwc io.ReadWriteCloser, err error) {
+	f, err := os.OpenFile(name, syscall.O_RDWR|syscall.O_NOCTTY|syscall.O_NONBLOCK, 0666)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err != nil && f != nil {
+			f.Close()
+		}
+	}()
+
+	fd := f.Fd()
+
+	t := &syscall.Termios{}
+
+	// Fetch old flags
+	if _, _, errno := syscall.Syscall6(
+		syscall.SYS_IOCTL,
+		uintptr(fd),
+		uintptr(syscall.TIOCGETA),
+		uintptr(unsafe.Pointer(t)),
+		0,
+		0,
+		0,
+	); errno != 0 {
+		return nil, errno
+	}
+
+	// These are mostly torn out of the pyserial implementation
+	// Clear char-size, stop-bits, even and odd parity as well as xon/xoff
+	t.Cflag &^= (syscall.CSIZE | syscall.CSTOPB | syscall.PARENB | syscall.PARODD | syscall.IXANY | syscall.IXON | syscall.IXOFF)
+	// Ignore modem control lines, enable receiver, 8-bit char-size
+	t.Cflag |= (syscall.CLOCAL | syscall.CREAD | syscall.CS8)
+	// Disable canonical mode, echo, echo erase, echo kill, echo newline, signal on INTR/QUIT/SUSP/DSUSP characters, impl. defined input processing
+	t.Lflag &^= (syscall.ICANON | syscall.ECHO | syscall.ECHOE | syscall.ECHOK | syscall.ECHONL | syscall.ISIG | syscall.IEXTEN)
+	// Disable impl. defined output processing
+	t.Oflag &^= (syscall.OPOST)
+	// Disable parity error marking, input parity checking, strip eighth bit, NL->CR translation, carriage-return ignore, CR->NL translation, break ignore
+	t.Iflag &^= (syscall.PARMRK | syscall.INPCK | syscall.ISTRIP | syscall.INLCR | syscall.IGNCR | syscall.ICRNL | syscall.IGNBRK)
+
+	// Minimum bytes for read
+	t.Cc[syscall.VMIN] = 1
+	// Timeout in deciseconds
+	t.Cc[syscall.VTIME] = 30
```

Comment:
Aaaah, so much time has passed since I wrote this patch, so my memory might be a bit flaky..!

All the flags are pySerial defaults (including the proper darwin baudrate control, which puts no restriction on baud rate apart from hardware support), which was the only "things work" reference I had. Replication of the bug is not consistent, and can take over 30 minutes to occur in the CNC dry-runs. I have been running this patch since the pull request, occasionally in 4+ hour runs at 115200 baud, without the freeze occurring.

I unfortunately don't remember much of the testing, but I am pretty sure I tried setting VMIN - it could easily cause a lockup as the kernel would wait for more data - before I gave up and just traced all options set by pySerial (Having the CNC router stop randomly is not exactly a healthy thing for the machine, tools or people around it). I believe that pySerial has considerably more battle experience, and seeing linux and generic modes of glossarial didn't apply the same flags anyway, I thought it would be better to find a tested reference. I would have put the same flags on posix and linux, but I don't have any non-darwin setups to perform real-life tests on, and would rather not commit untested code...

I can try to reproduce it using only the vmin/vtime settings, but inconsistent reproduction means that it will take a few hours of testing to hit it, and my job leaves my spare-time limited and my mind exhausted... :/